api/proto/config/appconfig.proto

```diff
@@ -26,6 +26,47 @@ enum MetaDataType {
   MetaDataDriveMultipart = 3; // Process multipart MIME for application
 }
 
+// The information about a snapshot
+message AvailableSnapshot {
+  string id = 1;
+}
+
+// The snapshot information for an Application Instance
+message SnapshotConfig {
+  // activeSnapshot is the id of the snapshot which is currently used.
+  // It may be empty if no snapshot is used. Worth to note that the field is
+  // not used alone to trigger a rollback. It just indicates that the device
+  // is currently using a snapshot with the given id. To trigger a rollback, in
+  // addition to it, the rollbackCounter should be increased. It is done to
+  // avoid unnecessary rollbacks when the same activeSnapshot is sent several
+  // times (we cannot relay on EVE to perform rollbacks after the first time
+  // it receives the corresponding message, so it's a regular case to send the
+  // same value several times). Also it is used to perform a rollback to the
+  // same snapshot several times.
+  string activeSnapshot = 1;
+  // rollbackCounter is necessary in addition to the activeSnapshot to request
+  // EVE to trigger a rollback. The rollbackCounter is incremented when a
+  // snapshot is used for a rollback. It should not be decreased.
+  uint32 rollbackCounter = 2;
+  // maxSnapshots is the maximum number of snapshots that can be stored for the
+  // application instance.
+  // If the number of snapshots exceeds the maxSnapshots, the oldest snapshot
+  // will be deleted.
+  uint32 maxSnapshots = 3;
+  // availableSnapshots is the list of available snapshots for the application
+  // instance.
+  // It is used to synchronize the list of snapshots between the controller and
+  // the device. This can be used to create a new snapshot or to delete an
+  // existing one. To create a new snapshot, the controller should send a
```

Comment:
This "create" sounds like the controller can decide to ask for a snapshot at any time, but that is for later (and requires a bunch of interaction with not only the guest VM but also with the applications running in the guest VM to flush any buffers to storage to ensure the snapshot is consistent from the applications perspective.

Given that we are not doing that in phase 1 and instead are only doing creation of snapshots as part of an update (and associated restart) of the app instance, does it make sense to have this? 
And who allocates the id for the snapshot which is taken as part of the application update?
I think we had talked about EVE-OS picking a unique ID using a timestamp, which implies that the id is (or includes) a timestamp.

Reply:
About the IDs.
I assume that if the controller triggers the snapshot, it can also provide the ID for it. EVE takes this ID and uses it (or uses it as a hint for a locally generated ID).
Meanwhile, EVE can assign an ID if the snapshot is created automatically on the device (up to future implementations).
In any case, the real ID of the snapshot (whether it was taken from the controller or generated locally) is reported by EVE as a part of the `ZInfoApp.snapshots` field.

Yep, we do not support snapshot creation at any moment. Nevertheless, the proposed API can be used universally on phase 1 and later.

Reply:
I don't think we want EVE to generate UUIDs, because they might collide in the controller (who knows what the scale might be, plus hardware with poor sources of random numbers.)

Having the controller provide a base UUID and have EVE-OS augment those with e.g., a timestamp or a local monotonically increasing number and using the combination of those means that we will have unique ways to identify the snapshots in the controller.

Reply:
The controller can store and identify the snapshot internally by a tuple of VM UUID and the snapshot ID. Snapshots do not without a VM in any case. So we need these IDs to be unique only in the scope of a VM.
@xyuria-zededa, what do you think about it?